libfwupd/fwupd-client-sync.c

```diff
@@ -2524,3 +2524,109 @@ fwupd_client_upload_bytes(FwupdClient *self,
 	}
 	return g_steal_pointer(&helper->bytes);
 }
+
+static void
+fwupd_client_emulate_load_cb(GObject *source, GAsyncResult *res, gpointer user_data)
+{
+	FwupdClientHelper *helper = (FwupdClientHelper *)user_data;
+	helper->ret = fwupd_client_emulate_load_finish(FWUPD_CLIENT(source), res, &helper->error);
+	g_main_loop_quit(helper->loop);
+}
+
+/**
+ * fwupd_client_emulate_load
+ * @self: a #FwupdClient
+ * @data: archive data of JSON files
+ * @cancellable: (nullable): optional #GCancellable
+ * @error: (nullable): optional return location for an error
+ *
+ * Loads an emulated device into the daemon backend that has the phases set by the JSON data,
+ * for instance, having one USB device emulated for the bootloader and another emulated for the
+ * runtime interface.
+ *
+ * Returns: %TRUE for success
+ *
+ * Since: 1.8.11
+ **/
+gboolean
+fwupd_client_emulate_load(FwupdClient *self,
+			  GBytes *data,
+			  GCancellable *cancellable,
+			  GError **error)
+{
+	g_autoptr(FwupdClientHelper) helper = NULL;
+
+	g_return_val_if_fail(FWUPD_IS_CLIENT(self), FALSE);
+	g_return_val_if_fail(data != NULL, FALSE);
+	g_return_val_if_fail(cancellable == NULL || G_IS_CANCELLABLE(cancellable), FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	/* connect */
+	if (!fwupd_client_connect(self, cancellable, error))
+		return FALSE;
+
+	/* call async version and run loop until complete */
+	helper = fwupd_client_helper_new(self);
+	fwupd_client_emulate_load_async(self,
+					data,
+					cancellable,
+					fwupd_client_emulate_load_cb,
+					helper);
+	g_main_loop_run(helper->loop);
+	if (!helper->ret) {
+		g_propagate_error(error, g_steal_pointer(&helper->error));
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static void
+fwupd_client_emulate_save_cb(GObject *source, GAsyncResult *res, gpointer user_data)
+{
+	FwupdClientHelper *helper = (FwupdClientHelper *)user_data;
+	helper->bytes = fwupd_client_emulate_save_finish(FWUPD_CLIENT(source), res, &helper->error);
+	g_main_loop_quit(helper->loop);
+}
+
+/**
+ * fwupd_client_emulate_save:
+ * @self: a #FwupdClient
+ * @cancellable: (nullable): optional #GCancellable
+ * @error: (nullable): optional return location for an error
+ *
+ * Gets the captured data from all filtered devices for all recorded phases. The data is returned
+ * in an archive (typically ZIP encoded) with JSON entries.
```

Comment:
what could cause it not to be zip encoded?

Reply:
Do you think we should enforce .zip? It's certainly a well understood format, and it just adds to the support matrix to have other formats. On the other hand, supporting .can would allow us to include the emulation in the firmware binary, although I'm not sure how insane that is.

Reply:
Ohh I see -- this is from the daemon, i.e. we know the format. Fixed.